core/src/test/java/org/apache/iceberg/TestSequenceNumberForV2Table.java

```diff
@@ -0,0 +1,299 @@
+    table.updateProperties()
+        .set(TableProperties.MANIFEST_MIN_MERGE_COUNT, "1")
+        .commit();
+    table.newAppend().appendFile(FILE_A).commit();
+    Assert.assertEquals(1, table.currentSnapshot().sequenceNumber());
+
+    table.newAppend().appendFile(FILE_B).commit();
+    Assert.assertEquals(2, table.currentSnapshot().sequenceNumber());
+
+    ManifestFile manifestFile = writeManifest(FILE_C, FILE_D);
+    table.newAppend().appendManifest(manifestFile).commit();
+    Assert.assertEquals(3, table.currentSnapshot().sequenceNumber());
+
+    manifestFile = table.currentSnapshot().manifests().get(0);
+
+    Assert.assertEquals("the sequence number of manifest should be 3", 3, manifestFile.sequenceNumber());
+
+    for (ManifestEntry entry : ManifestFiles.read(manifestFile,
+        table.io(), table.ops().current().specsById()).entries()) {
+      if (entry.file().path().equals(FILE_A.path())) {
+        Assert.assertEquals("the sequence number of data file should be 1", 1, entry.sequenceNumber().longValue());
+      }
+
+      if (entry.file().path().equals(FILE_B.path())) {
+        Assert.assertEquals("the sequence number of data file should be 2", 2, entry.sequenceNumber().longValue());
+      }
+
+      if (entry.file().path().equals(FILE_C.path()) || entry.file().path().equals(FILE_D.path())) {
+        Assert.assertEquals("the sequence number of data file should be 3", 3, entry.sequenceNumber().longValue());
+      }
+    }
+  }
+
+  @Test
+  public void testSequenceNumberForRewrite() {
+    table.newFastAppend().appendFile(FILE_A).commit();
+    table.newFastAppend().appendFile(FILE_B).commit();
+    Assert.assertEquals(2, table.currentSnapshot().sequenceNumber());
+
+    table.rewriteManifests().clusterBy(file -> "").commit();
+    Assert.assertEquals("the sequence number of snapshot should be 3",
+        3, table.currentSnapshot().sequenceNumber());
+
+    ManifestFile newManifest = table.currentSnapshot().manifests().get(0);
+    Assert.assertEquals("the sequence number of manifest should be 3",
+        3, newManifest.sequenceNumber());
+
+    for (ManifestEntry entry : ManifestFiles.read(newManifest,
+        table.io(), table.ops().current().specsById()).entries()) {
+      if (entry.file().path().equals(FILE_A.path())) {
+        Assert.assertEquals("the sequence number of data file should be 1", 1, entry.sequenceNumber().longValue());
+      }
+
+      if (entry.file().path().equals(FILE_B.path())) {
+        Assert.assertEquals("the sequence number of data file should be 1", 2, entry.sequenceNumber().longValue());
+      }
+    }
+  }
+
+  @Test
+  public void testCommitConflict() {
+    Transaction txn = table.newTransaction();
+
+    txn.newFastAppend().appendFile(FILE_A).apply();
+    table.newFastAppend().appendFile(FILE_B).commit();
+
+    AssertHelpers.assertThrows("Should failed due to conflict",
+        IllegalStateException.class, "last operation has not committed", txn::commitTransaction);
```

Comment:
I think there is some confusion about what this is doing. This `IllegalStateException` indicates that one of the operations in a transaction was not committed, not that there was a conflict committing to the table. If you want to validate a conflict, then you need to set the number of retries to 0 so that the `CommitFailedException` is thrown.

Most tests that validate a conflict call apply on the first change, then call commit on the conflicting change, and then call commit on the first change. Retry allows the commit to succeed, but the result of the commit will be different. In this case, you'd have a higher sequence number.

Reply:
Updated.